File: hupu/api/live.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Time    : 2018/1/27 上午3:58
# @Author  : wudizhangzhi
import datetime
import json
import time
from random import choice

from hupu.api import logger
from hupu.api.base import Base
from hupu.messages.messages import Game

log = logger.getLogger(__name__)


class LiveMinxin(Base):
    def _get_gamesobject_list(self, games_json):
        """
        :param games_json: 
        :return: 
        """
        today = datetime.datetime.today().strftime('%Y%m%d')

        games = games_json['result']['games']
        games_today = [i for i in games if i['day'] == today]
        games_today = games_today[0]
        # rank_url = games_today['rank_url']
        game_datas = games_today['data']

        game_object_list = []
        for _game_data in game_datas:
            gid = _game_data['gid']
            if len(gid) > 6:
                continue
            g = Game(_game_data)
            game_object_list.append(g)
        return game_object_list

    def _getGames(self, gametype=None):
        """
        获取比赛列表
        """
        if gametype:
            assert gametype.lower() in ['nba', 'cba']
        else:
            gametype = 'nba'
        return self.sess.get(
            url='https://games.mobileapi.hupu.com/1/{}/{}/getGames'.format(self.api_version, gametype),
            params={
                'crt': int(time.time() * 1000),
                'night': 0,
                'channel': 'myapp',
                'client': self.client,
                'time_zone': 'Asia/Shanghai',
                'android_id': self.android_id,
            }
        )

    def getGames(self, gametype=None):
        r = self._getGames(gametype).json()
        log.debug('获取比赛列表: {}'.format(r))
        return self._get_gamesobject_list(r)

    def getPlaybyplay(self, gid):
        """
        获取比赛直播信息
        :param gid: 
        :return: 
        """
        return self.sess.get(
            url='https://games.mobileapi.hupu.com/1/{}/room/getPlaybyplay'.format(self.api_version),
            params={
                'gid': gid,
                'crt': int(time.time() * 1000),
                'lid': '1',
                'roomid': '-1',
                'tag': 'nba',  # 默认nba
                'night': 0,
                'channel': 'myapp',
                'client': self.client,
                'time_zone': 'Asia/Shanghai',
                'android_id': self.android_id,
                'entrance': '-1',
            },
        )

    def getInit(self):
        return self.sess.get(
            url='https://games.mobileapi.hupu.com/1/{}/status/init'.format(self.api_version),
            params={
                'dv': '5.7.79',
                'crt': int(time.time() * 1000),
                'tag': 'nba',  # 默认nba
                'night': 0,
                'channel': 'myapp',
                'client': self.client,
                'time_zone': 'Asia/Shanghai',
                'android_id': self.android_id,
            },
        )

    def getIpAdress(self):
        host = port = None

        try:
            r_json = self.getInit().json()
            log.debug('ip地址获取: {}'.format(json.dumps(r_json, indent=2)))
            ip_adress_list = choice(r_json['result']['redirector'])
            log.debug('ip地址使用: {}'.format(ip_adress_list))
            tmp = ip_adress_list.split(':')
            host = tmp[0]
            port = int(tmp[1])
        except:
            pass
        return host, port

File: gohupu/api/live.go
package api

import (
	"fmt"

	"github.com/go-resty/resty/v2"
	"github.com/wudizhangzhi/HupuApp"
)

type GameType string

const (
	NBA GameType = "nba"
	CBA GameType = "cba"
)

func (g GameType) String() string {
	switch g {
	case NBA:
		return "nba"
	case CBA:
		return "cba"
	default:
		return "nba"
	}
}

func init() {

}

// 状态初始化，获取基本信息接口
func GetInitInfo() (*resty.Response, error) {
	params := map[string]string{
		"div":        "5.7.79",
		"crt":        fmt.Sprint(HupuApp.GetTimestamp()),
		"tag":        "nba",
		"night":      "0",
		"channel":    "myapp",
		"client":     HttpSession.IMEI,
		"time_zone":  "Asia/Shanghai",
		"android_id": HttpSession.AndroidId,
	}
	return HttpSession.Request("GET", HupuApp.API_STATUS_INIT, nil, params)
}

// 获取比赛直播信息
func GetPlayByPlay(gid int) (*resty.Response, error) {
	params := map[string]string{
		"gid":        fmt.Sprint(gid),
		"lid":        "1",
		"roomid":     "-1",
		"entrance":   "-1",
		"tag":        "nba",
		"channel":    "myapp",
		"night":      "0",
		"crt":        fmt.Sprint(HupuApp.GetTimestamp()),
		"client":     HttpSession.IMEI,
		"time_zone":  "Asia/Shanghai",
		"android_id": HttpSession.AndroidId,
	}
	return HttpSession.Request("GET", HupuApp.API_GET_PLAY_BY_PLAY, nil, params)
}

// 获取比赛直播信息
func GetLiveActivityKey(matchId string) (*resty.Response, error) {
	params := map[string]string{
		"competitionType": "basketball",
		"matchId":         matchId,
		"channel":         "hupuupdate",
		"night":           "0",
		"crt":             fmt.Sprint(HupuApp.GetTimestamp()),
		"_imei":           HttpSession.IMEI,
		"time_zone":       "Asia/Shanghai",
		"android_id":      HttpSession.AndroidId,
	}
	return HttpSession.Request("GET", HupuApp.API_LIVE_QUERY_LIVE_ACTIVITY_KEY, nil, params)
}

// 获取直播内容接口
func GetLiveMsgList(matchId string, liveActivityKeyStr string, commentId string) (*resty.Response, error) {
	params := map[string]string{
		"competitionType":    "basketball",
		"matchId":            matchId,
		"liveActivityKeyStr": liveActivityKeyStr,
		"channel":            "hupuupdate",
		"night":              "0",
		"crt":                fmt.Sprint(HupuApp.GetTimestamp()),
		"_imei":              HttpSession.IMEI,
		"time_zone":          "Asia/Shanghai",
		"android_id":         HttpSession.AndroidId,
	}
	if commentId != "" {
		params["commentId"] = commentId
	}
	return HttpSession.Request("GET", HupuApp.API_LIVE_QUERY_LIVE_TEXT_LIST, nil, params)
}

// 获取比赛日程列表
func GetScheduleList(gametype GameType, coursors ...string) (*resty.Response, error) {
	params := map[string]string{
		"competitionTag": gametype.String(),
		"night":          "0",
		"V":              "7.5.59.01043",
		"channel":        "hupuupdate",
		"crt":            fmt.Sprint(HupuApp.GetTimestamp()),
		"_imei":          HttpSession.IMEI,
		"time_zone":      "Asia/Shanghai",
		"android_id":     HttpSession.AndroidId,
	}
	if len(coursors) > 0 {
		params["coursor"] = coursors[0]
	}
	return HttpSession.Request("GET", HupuApp.API_SCHEDULE_LIST, nil, params)
}

// 根据比赛id获取比赛信息
func GetSingleMatch(matchId string) (*resty.Response, error) {
	params := map[string]string{
		"matchId":    matchId,
		"night":      "0",
		"V":          "7.5.59.01043",
		"channel":    "hupuupdate",
		"crt":        fmt.Sprint(HupuApp.GetTimestamp()),
		"_imei":      HttpSession.IMEI,
		"time_zone":  "Asia/Shanghai",
		"android_id": HttpSession.AndroidId,
	}
	return HttpSession.Request("GET", HupuApp.API_SINGLE_MATCH, nil, params)
}
